test/gtest/uct/test_peer_failure.cc

```diff
@@ -140,10 +141,17 @@ void test_uct_peer_failure::set_am_handlers()
 ucs_status_t test_uct_peer_failure::send_am(int index)
 {
     ucs_status_t status;
+
     while ((status = uct_ep_am_short(m_sender->ep(index), 0, 0, NULL, 0)) ==
            UCS_ERR_NO_RESOURCE) {
         progress();
-    };
+
+        status = ep_status(index);
+        if (status != UCS_OK) {
+            break;
+        }
```

Comment:
why needed since we are not supposed to call send_am on failed ep?

Reply:
We do progress(), so the failure on EP could be detected after progress

Reply:
in which line?

Reply:
we do progress() on `test/gtest/uct/test_peer_failure.cc:147`

Reply:
why is it important to return ep_status from this function, and can't just return the result of uct_ep_am_short?

Reply:
uct_ep_am_short will return UCS_OK, even if error detected

Reply:
but aren't we skipping send_am if error is detected?

Reply:
> but aren't we skipping send_am if error is detected?

yes, we do it
but the error can be detected during doing progress in `send_am`

Reply:
so maybe move checking failed eps set logic to send_am while loop (and remove ep_status wrapper), and also remove check before calling send_am?

Reply:
> so maybe move checking failed eps set logic to send_am while loop (and remove ep_status wrapper), and also remove check before calling send_am?

but we should if the EP failed before doing uct_ep_am_short.
also, we should return the status instead of `UCS_OK` from `send_am` to stop doing futher `send_am`.

wdyt about the following?
```
do {
        progress();

        status = ep_status(index);
        if (status != UCS_OK) {
            break;
         }
} while (uct_ep_am_short(m_sender->ep(index), 0, 0, NULL, 0) == UCS_ERR_NO_RESOURCE);

return status;
```

Reply:
(suggestion only)
```
for (;;) {
   /* if endpoint has failed, return error and avoid calling send
   it = find(..)
   if (!= end) 
     return it->second
   status = am_send()
  progress
}
```

Reply:
done